In [1]:
from stockbot.backtest_bot import *
import pandas as pd
from pathlib import Path
from datetime import datetime

##### Back end configurations for the **Backtest Bot**

- Set the data_folder
- Change the YAML configuration file to the file containing strategies you want to test. 

In [2]:
data_folder = Path('.\\config')
yaml_config = 'extended_backtesting.yaml'
cfg = data_folder / yaml_config


##### Load the config file into the Backtest class and run the analysis

In [3]:
backtest = Backtest(cfg)
results = backtest.run()


In [4]:
# Get Performance Metrics for each strategy contained in the YAML Config file. 
backtest.get_performance_metrics()

{'ma_crossover_aapl': {'total_return': np.float64(-5.407626718222569),
  'annual_return': np.float64(-1.132523495178972),
  'avg_daily_return': np.float64(0.0130554123418317),
  'return_std': np.float64(1.8752821420905876),
  'drawdown_metrics': {'max_drawdown': np.float64(-0.5689129716967678),
   'avg_drawdown': np.float64(-0.34631286331158223),
   'max_drawdown_duration': np.int64(1061),
   'avg_drawdown_duration': np.float64(487.96990541702496),
   'drawdown_frequency': np.float64(0.9455284552845529)},
  'trade_metrics': {'total_trades': 27,
   'win_rate': np.float64(0.38461538461538464),
   'avg_trade_return': np.float64(0.0024952349476966306),
   'best_trade': np.float64(0.4463752121562927),
   'worst_trade': np.float64(-0.17583247946369018),
   'avg_winning_trade': np.float64(0.1114297704943192),
   'avg_losing_trade': np.float64(-0.06558884976894247),
   'profit_factor': np.float64(1.061820824794019)},
  'risk_metrics': {'sharpe_ratio': np.float64(0.04399258975285559),
   'sorti

#### Results Breakdown

##### 1. For the *Bollinger Portfolio* (best performer):

**Return Metrics**:

- Total Return: +19.87% (total profit)
- Annual Return: +3.79% per year
- Average Daily Return: +0.017% per day
- Return Standard Deviation: 0.69% daily volatility


**Trade Performance**:


* Win Rate: 74.47% (percentage of profitable trades)
* Total Trades: 142 trades
* Best Trade: +11.98%
* Worst Trade: -5.41%
* Profit Factor: 3.79 (gross profits divided by gross losses - very good)


**Risk Metrics**:


* Sharpe Ratio: 0.21 (risk-adjusted returns - positive but low)
* Maximum Drawdown: -9.24% (biggest peak-to-trough decline)
* Value at Risk (95%): 0%
* Volatility: 10.89% annualized

##### 2. For the *RSI Tech Basket* (moderate performer):

**Return Metrics**:


- Total Return: +3.66%
- Annual Return: +0.74%
- Average Daily Return: +0.008%


**Trade Performance**:


* Win Rate: 69.52%
* Total Trades: 188 trades
* Best Trade: +9.31%
* Worst Trade: -7.56%
* Profit Factor: 2.42 (good)

##### 3. For the MA Crossover AAPL (worst performer):

**Return Metrics**:


* Total Return: -6.77% (loss)
* Annual Return: -1.43%
* Average Daily Return: +0.012%


**Trade Performance**:


- Win Rate: 36% (poor)
- Total Trades: 26 trades
- Best Trade: +44.64%
- Worst Trade: -17.58%
- Profit Factor: 1.06 (barely profitable)

The Bollinger Portfolio strategy performed best with:

         Highest total return
         Best win rate
         Best profit factor
         Smallest maximum drawdown
         Most consistent performance metrics

In [5]:
backtest._get_all_symbols()

['NVDA', 'NIO', 'MSFT', 'GOOGL', 'PLTR', 'AAPL']

In [6]:
backtest.strategies

{'ma_crossover_aapl': <stockbot.Strategy.MovingAverageCrossover at 0x20628b212b0>,
 'rsi_tech_basket': <stockbot.Strategy.RSIStrategy at 0x20628b21010>,
 'bollinger_portfolio': <stockbot.Strategy.BollingerBandsStrategy at 0x20628b20c20>,
 'ichimoku_test': <stockbot.Strategy.IchimokuStrategy at 0x20628b20ec0>}

In [7]:
backtest.config

{'data': {'universe': {'type': 'static', 'symbols': ['AAPL', 'MSFT', 'GOOGL']},
  'parameters': {'interval': '1d',
   'history_start': '2020-01-01',
   'add_indicators': True,
   'cache_data': True,
   'adjustment': 'split+div'}},
 'strategies': {'ma_crossover_aapl': {'type': 'MovingAverageCrossover',
   'enabled': True,
   'symbols': ['AAPL'],
   'parameters': {'short_window': 20, 'long_window': 50}},
  'rsi_tech_basket': {'type': 'RSIStrategy',
   'enabled': True,
   'symbols': ['MSFT', 'GOOGL'],
   'parameters': {'period': 14, 'overbought': 70, 'oversold': 30}},
  'bollinger_portfolio': {'type': 'BollingerBandsStrategy',
   'enabled': True,
   'symbols': ['AAPL', 'MSFT', 'GOOGL'],
   'parameters': {'period': 20, 'std_dev': 2, 'mean_reversion': True}},
  'ichimoku_test': {'type': 'IchimokuStrategy',
   'enabled': True,
   'symbols': ['NVDA', 'PLTR', 'NIO'],
   'parameters': {'conversion_period': 9,
    'base_period': 26,
    'span_period': 52}}},
 'portfolio': {'initial_capital': 100